k8scrdclient/k8scrdclient.go

```diff
@@ -120,24 +122,26 @@ func (c *CRDClient) ensureCreated(ctx context.Context, crd *apiextensionsv1beta1
 	return nil
 }
 
-// ensureStatusSubresourceCreated ensures if the CRD has a status subresource
-// it is created. This is needed if a previous version of the CRD without the
-// status subresource is present.
-func (c *CRDClient) ensureStatusSubresourceCreated(ctx context.Context, crd *apiextensionsv1beta1.CustomResourceDefinition, b backoff.Interface) error {
-	if crd.Spec.Subresources == nil || crd.Spec.Subresources.Status == nil {
-		// Nothing to do.
-		return nil
-	}
-
+// ensureUpdated ensures if the CRD changed it is updated accordingly. This is
+// needed if e.g. a previous version of the CRD without the status subresource
+// is present where it should actually be set. Another example would be the CRD
+// apiversion changing, which tends to happen every now and then over the
+// runtime object lifecycle and community adoption.
+func (c *CRDClient) ensureUpdated(ctx context.Context, desired *apiextensionsv1beta1.CustomResourceDefinition, b backoff.Interface) error {
 	o := func() error {
-		manifest, err := c.k8sExtClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(crd.Name, metav1.GetOptions{})
+		current, err := c.k8sExtClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(desired.Name, metav1.GetOptions{})
```

Comment:
Really dumb question:
This client is configured with a `schema` I believe - are we sure that it will actually return the accurate crd version? It could read convert to the same version as the `desired` version here because of how the client is configured.

Far fetched but could be the case.

Reply:
Discussed with Marcel. Good point about the API Version of the CRD itself. Actual goal here was rather about the Schema Validation versions if that follows. We need to support some migration logic in the client here and deploy new operator versions when making the step from `apiextensions.k8s.io/v1beta1` to `apiextensions.k8s.io/v1`.